core/src/main/java/org/apache/iceberg/jdbc/JdbcTableOperations.java

```diff
@@ -138,7 +139,9 @@ public void doCommit(TableMetadata base, TableMetadata metadata) {
       throw new UncheckedSQLException(e, "Database warning");
     } catch (SQLException e) {
       // SQLite doesn't set SQLState or throw SQLIntegrityConstraintViolationException
-      if (e.getMessage() != null && e.getMessage().contains("constraint failed")) {
+      // Postgres doesn't throw SQLIntegrityConstraintViolationException
+      if (e.getMessage() != null
+          && (e.getMessage().contains("constraint failed") || "23505".equals(e.getSQLState()))) {
```

Comment:
nit: I think e.getMessage() != null is no need to be checked for "23505". How about
```
if ("23505".equals(e.getSQLState())
    || (e.getMessage() != null && e.getMessage().contains("constraint failed"))) {
  throw new AlreadyExistsException("Table already exists: %s", tableIdentifier);
}
```